cfme/middleware/deployment.py

```diff
@@ -59,24 +67,38 @@ def __init__(self, name, provider=None, **kwargs):
 
     @classmethod
     def deployments(cls, provider):
-        sel.force_navigate('middleware_deployments')
         deployments = []
+        if provider:
+            # if provider instance is provided try to navigate provider's deployments page
+            # if no deployments are registered in provider, returns empty list
+            if provider.load_all_provider_deployments():
+                return deployments
+        else:
+            # if provider instance is not provided then navigates  to all deployments page
+            sel.force_navigate('middleware_deployments')
         for row in list_tbl.rows():
-            deployments.append(MiddlewareDeployment(
-                provider=provider, name=row.deployment_name.text, server=row.server.text))
+            deployments.append(MiddlewareDeployment(name=row.deployment_name.text,
+                                                    server=row.server.text))
         return deployments
 
     @classmethod
-    def deployments_in_db(cls, server=None):
+    def deployments_in_db(cls, server=None, provider=None):
         deployments = []
-        rows = _db_select_query(server=server).all()
+        rows = _db_select_query(server=server, provider=(provider.name if provider else None)).all()
         for deployment in rows:
-            deployments.append(Deployment(deployment[0], deployment[1], Path(deployment[2])))
+            deployments.append(MiddlewareDeployment(id=deployment[0], name=deployment[1],
+                                                    server=deployment[3], provider=deployment[4]))
         return deployments
 
     @classmethod
     def deployments_in_mgmt(cls, provider):
-        return provider.mgmt.list_server_deployment()
+        deployments = []
+        rows = provider.mgmt.list_server_deployment()
+        for deployment in rows:
+            deployments.append(MiddlewareDeployment(provider=provider.name,
+                    id=deployment.id, name=deployment.name,
+                    server=re.sub(r'(~~$)', '', deployment.path.resource[0])))
```

Comment:
I guess the parentheses are not necessary in the regular expression as you don't use the captured string in the replacement string.